File: app/services/completed_loan_services/tab_income.rb
module CompletedLoanServices
  class TabIncome
    attr_accessor :borrower, :current_employment,
                  :previous_employment, :secondary_borrower

    def initialize(args)
      @borrower = args[:borrower]
      @current_employment = args[:current_employment]
      @previous_employment = args[:previous_employment]
      @secondary_borrower = args[:secondary_borrower]
    end

    def call
      return false unless current_employment
      return false unless employment_completed?(current_employment)
      return false unless current_employment_duration_valid?
      return false unless borrower.gross_income

      unless secondary_borrower.nil?
        return false unless secondary_borrower.current_employment
        return false unless employment_completed?(secondary_borrower.current_employment)
        return false unless co_current_employment_duration_valid?
        return false unless secondary_borrower.gross_income
      end

      true
    end

    def employment_completed?(employment)
      return false unless employment.employer_name.present?
      return false unless employment.address.present?
      return false unless employment.employer_contact_name.present?
      return false unless employment.employer_contact_number.present?
      return false unless employment.current_salary.present?
      return false unless employment.job_title.present?
      return false unless employment.pay_frequency.present?
      return false unless employment.duration.present?

      true
    end

    def current_employment_duration_valid?
      current_employment.duration >= 2 || (current_employment.duration < 2 && previous_employment_completed?(previous_employment))
    end

    def co_current_employment_duration_valid?
      secondary_borrower.current_employment.duration >= 2 || (secondary_borrower.current_employment.duration < 2 && previous_employment_completed?(secondary_borrower.previous_employment))
    end

    def previous_employment_completed?(prev_employment)
      return false unless prev_employment
      return false unless prev_employment.employer_name.present?
      return false unless prev_employment.job_title.present?
      return false unless prev_employment.duration.present?
      return false unless prev_employment.monthly_income.present?

      true
    end
  end
end

File: client/assets/javascripts/mixins/CompletedLoanMixins/TabIncome.js
var TabIncome = {
  /**
   * Check tab income is completed or not
   * @type {boolean}
   */
  incomeCompleted: function(loan){
    var borrower = loan.borrower;
    var secondaryBorrower = loan.secondary_borrower;
    var currentEmployment = borrower.current_employment;
    var previousEmployment = borrower.previous_employment;

    if(this.objectFieldNotValid(currentEmployment))
      return false;

    if(!this.employmentCompleted(currentEmployment))
      return false;

    if(!this.currentEmploymentDurationValid(currentEmployment, previousEmployment))
      return false;

    if(this.valueFieldNotValid(borrower.gross_income))
      return false;

    if (!this.objectFieldNotValid(secondaryBorrower)){

      if(this.objectFieldNotValid(secondaryBorrower.current_employment))
        return false;

      if(!this.employmentCompleted(secondaryBorrower.current_employment))
        return false;

      if(!this.currentEmploymentDurationValid(secondaryBorrower.current_employment, secondaryBorrower.previous_employment))
        return false;

      if(this.valueFieldNotValid(secondaryBorrower.gross_income))
        return false;
    }

    return true;
  },

  employmentCompleted: function(employment){
    if (this.valueFieldNotValid(employment.employer_name))
      return false;

    if (this.valueFieldNotValid(employment.address))
      return false;

    // if (this.valueFieldNotValid(employment.employer_contact_name))
    //   return false;

    // if (this.valueFieldNotValid(employment.employer_contact_number))
    //   return false;

    if (this.valueFieldNotValid(employment.current_salary))
      return false;

    if (this.valueFieldNotValid(employment.job_title))
      return false;

    if (this.valueFieldNotValid(employment.pay_frequency))
      return false;

    if (this.valueFieldNotValid(employment.duration))
      return false;

    return true;
  },

  currentEmploymentDurationValid: function(currentEmployment, previousEmployment){
    return (currentEmployment.duration >= 2 || (currentEmployment.duration < 2 && this.previousEmploymentCompleted(previousEmployment)));
  },

  previousEmploymentCompleted: function(prevEmployment){
    if (this.objectFieldNotValid(prevEmployment))
      return false;

    if (this.valueFieldNotValid(prevEmployment.employer_name))
      return false;

    if (this.valueFieldNotValid(prevEmployment.job_title))
      return false;

    if (this.valueFieldNotValid(prevEmployment.duration))
      return false;

    if (this.valueFieldNotValid(prevEmployment.monthly_income))
      return false;

    return true;
  },

  valueFieldNotValid: function(field){
    if (field === undefined || field === null || field === "")
      return true;

    return false;
  },

  objectFieldNotValid: function(obj){
    if(obj === undefined || obj === null)
      return true;

    return false;
  },
}

module.exports = TabIncome;